README.MD
# RDR Spreadsheet Engineering

[![clasp](docs/img/built-with-clasp-4285f4.svg)](https://github.com/google/clasp)

> Brace yourself: [Why software projects take longer than you think – a statistical model](https://erikbern.com/2019/04/15/why-software-projects-take-longer-than-you-think-a-statistical-model.html)

—

## TOC

+ [Intro](#intro)
+ [Features](#features)
+ [Quick Start](#quick-start)
+ [Changelog](#changelog)
+ [Documentation](#documentation)

—

## Intro

This is the master repository for RDR's spreadsheet automation & data pipeline development.

![The development environment](docs/img/dev_environment.jpg)
(*The development environment. Clockwise: VSCode+clasp; Google Apps Script; GitHub; Spreadsheet Output on Drive*)

The ~~current~~ former scope of the Spreadsheet development was to automate the generating ("casting") of spreadsheets for company-level data collection (DC) and for scoring (SC) with as few manual effort as possible. In addition we wanted to enable layout-independent querying of input data (by ID vs. static Kartesian range system G5!B14:D19). The project has grown since it's inauguration in August 2019, mostly based on demand and feature discoveries during the 2019/20 Pilot, resulting in the introduction of additional modules (such as a scores aggregation or company feedback request forms) and functionalities.

The **current Scope** of the project is to complete an end-to-end data pipeline (spanning from previous Index' data to the final outputs from the current Index cycle).

## Features

Currently, the application's **core features** allow to auto-generate:

+ company-level input spreadsheets
+ company-level scoring spreadsheets
+ company-level feedback request spreadsheets ("forms")
+ summary scores spreadsheets
+ horizontal review spreadsheets
+ company level data stores (tidy results & scores data)

In addition, **additional modules** and functions allow to:

+ audit & maintain input sheet integrity (restoring named ranges with manipulating content)
+ update non-data parts (denominators, element level descriptions, formatting)
+ fetch Google Drive activity statistics for files and folders

**Current project  goals:**

+ [ ] migrate from rdresearch to data@rdr & dev@rdr
+ [ ] implement permission management
+ [ ] integrate & automate a step-wise company-level data flow starting with the previous Index' results
+ [ ] develop & test a management dashboard, which provides an overview of research progress and identifies anomalies and other prompts for manual action (i.e. upon the return of company feedback)
+ [ ] Connect Data Store Module to external DB & enable RESTful POST UPDATE([IDs])
+ [ ] migrate to `ES6` as soon as it is ~~available~~ **stable** for GAS (cf. [v8 runtime](https://www.labnol.org/es6-google-apps-script-v8-200206); [Google's overview](https://developers.google.com/apps-script/guides/v8-runtime))

**Current work-in progress:**

+ map folder structure
+ change range level permissions
+ change file level permission

—

## Environment

The environment utilizes [Google Apps Script](https://developers.google.com/apps-script/) (short: GAS]) (which is mostly a JavaScript flavor with G-Suite specific classes and methods), and the respective [Google Apps Script API](https://www.google.com/script/start/). Code is executed through the [browser](https://script.google.com/home/my) and runs online on Google's servers. Output is stored in Google Drive. 

While you *can edit* and maintain the code in Google's web editor, this is strongly discouraged. To enable proper versioning, [SSOT](https://en.wikipedia.org/wiki/Single_source_of_truth), code-sharing, collaboration, formatting consistency (`eslint` & `.editorconfig`) issue-tracking, and so on, it is recommended **to use Git, [`clasp`](https://github.com/google/clasp), and `VSCode` (for consistency and easier integration with NodeJS/npm). This setup enables you to edit code locally with a proper development environment, version your code in a git* repository, and deploy the code to GAS to execute online.

> **TL;DR:**
>
> + define parameters in JSON files (indicators, companies, research steps, Index parameters)
> + develop locally
> + deploy & run online

Following the initial proof-of-concept with G.W., the `project` has been migrated ~~to the `rdresearch` G-Suite account~~. In theory, you can share projects with other accounts in the same manner as you share any Google File. However, executing the code with any other account than `rdresearch` is not authorized currently (leads to error: `admin_policy_enforced`; probable scope: `https://www.googleapis.com/auth/script.external_request`). This will change soon by migrating the production environment from rdresearch@oti to dev@rdr.

## Conventions

+ Google Apps Script uses `*.gs` as the filetype, while regular Javascript uses `.js`. We can (*should*) stick to `.js`(or `.ts` for TypeScript) and ignore implications as clasp auto-converts between file types.

+ **IMPORTANT**: Keep in mind that the source code is stored in git. Any code changes you make in the online editor won't be fed back into git, unless you use `clasp pull` from your IDE (**danger!!!**) or copy/paste code from your browser into git. If you `push` \ `pull`, **all target folder's files will be overwritten**.

—

## Quick Start

1. log-in as `rdresearch`
2. open [My Projects](https://script.google.com/home/my) to get an overview of your GAS projects or go directly to the [`Spreadsheet Automation` project](https://script.google.com/a/opentechinstitute.org/d/1ZrUTGLLDXMZxkDB8BRaBpPb-p4ObTJrKI8FfJUN6cL10Iggc0TTalSC5)
3. `00_mainController.gs` should open automatically. If not, open the script. This is the **main interface** to config and run all **core** modules.
4. ~~edit `companyShortName` in alignment to the according `<company>.json` (see the [JSON folder](/json/))~~ (now vectorized; subset with `.slice(0,1)`)
5. double-check `outputFolderName` if you need a specific folder for your outputs (folder will be a subfolder of `2019 Back-end testing` owned by `rdresearch`; if you want to change the parent folder, you need to replace the folder **id**  of `rootFolderID`)
6. edit the `subset` variables in the `mainAllCompanies<Function>()` caller of the module you want to run (i.e. define whether you want to run only a subset of research steps)
7. in the sub-menu (grey), choose the function to run, and click play.
8. wait for the script to finish. When done, optionally hit `Ctrl + Enter` to inspect the log.
9. open the output folder in Google Drive (should be under `My Drive`) and inspect the output
10. if you created new data collection sheets: open `00_Test` spreadsheet and copy the DC Spreadsheet IDs to the

> If you created new data collection or scoring Spreadsheets, and you intend to continue working with them, make sure to grab the new file IDs from `00_Test` spreadsheet and update them for the respective `01_JSON_companies.js/<company>`, i.e. params

+ `companies.<company>.urlCurrentDataCollectionSheet`
+ `companies.<company>.urlCurrentCompanyScoringSheet`

Otherwise, you will have to update `=IMPORTRANGE(ID)` in the spreadsheets by hand.

> If you run into runtime issues (e.g. endless loop), you can inspect and **terminate** current executions under [My Executions](https://script.google.com/u/3/home/executions). The status in the online editor is just a front-end gimmick and is actually not at all connected to the execution itself, i.e. you can refresh or close the browser and/or start a new execution, and it will not impact the one already running.

—

## Changelog

> Moved to [**Changelog.md**](docs/CHANGELOG.md)

Current `cloc / Count Lines of Code` (2020-04-14):

### Summary

Language|files|blank|comment|code
:-------|-------:|-------:|-------:|-------:
JavaScript|47|1829|1089|5677
Markdown|1|20|0|53
--------|--------|--------|--------|--------
SUM:|48|1849|1089|5730

### Detailed

File|blank|comment|code
:-------|-------:|-------:|-------:
json/JSON_indicators.js|1|10|1145
src/11_inputSubModules.js|231|106|509
json/JSON_researchStepsPilot.js|0|0|463
src/23_scoringSubcomponents.js|184|88|392
src/72_repairSubcomponents.js|181|69|364
json/JSON_companies.js|2|2|287
src/42_feedbackHelper.js|90|32|232
src/33_aggregationHelper.js|69|40|163
src/11_inputMainProcess.js|74|43|130
00_mainController.js|76|86|117
src/99_SpreadsheetOperations.js|46|21|117
src/63a_dataStorageComponentsWide.js|49|29|108
src/999_drive_activity.js|58|65|100
src/999_driveActivityThesis.js|58|65|100
src/96_scoringFormulasHelper.js|33|4|100
src/22_scoringSingleStepProcess.js|44|13|97
src/63b_dataStorageComponentsLong.js|44|19|89
json/JSON_config.js|0|4|88
src/12_inputPermissions.js|43|35|86
src/41_feedbackInterface.js|48|17|80
src/10_inputInterface.js|43|18|69
src/62b_dataStorageSingleStepLong.js|38|15|68
src/61_dataStorangeInterface.js|36|18|64
src/30_aggregationMain.js|41|26|60
src/62a_dataStorageSingleStepWide.js|28|8|60
src/999_driveFilestructure.js|30|30|56
src/91_defaultSheets.js|15|1|54
src/72_repairBrokenREF.js|54|68|54
docs/CHANGELOG.md|20|0|53
src/40_feedbackMain.js|26|13|51
src/21_scoringInterface.js|20|8|37
src/71_repairInterface.js|20|9|34
src/95_formattingHelper.js|3|2|30
src/92_findItems.js|10|4|26
src/998_getSSfromFolder.js|8|3|25
src/20_scoringMain.js|18|12|24
src/32_aggregationCompanySummary.js|20|24|24
src/72_listBrokenREF.js|6|12|21
src/60_dataStorageLayer.js|16|24|21
src/70_repairSpreadsheets.js|14|15|20
src/99_importJSON.js|13|14|18
src/31_aggregationInterface.js|11|8|18
src/98_driveOperations.js|1|1|16
src/92_rangeNamingHelper.js|8|5|16
src/92_fileNamingHelper.js|6|1|14
src/93_dataHelpers.js|5|0|13
src/997_external_libraries_js.js|3|2|11
src/79_clearNamedRanges.js|5|0|6
--------|--------|--------|--------
SUM:|1849|1089|5730

---

## Detailed Documentation

> To keep things readable, the documentation is maintained in a separate document.

+ [**Documentation**](docs/documentation.MD)


docs/CHANGELOG.md
# Changelog

> Keep track of new developments here, document your progress, also explicitly indicate **breaking** changes

## 2020 April

+ Breaking: Project setup migrated
  + dev@ owns project
  + data@ owns data & outputs
  + users need to use their own accounts to deploy & execute modules
  + login and admin tasks reflected in [setup.md](/docs/setup.md)
+ Main readme & changelog overhauled updated (rolling);
+ git branches cleanup
  + development for main feature testing
  + temporary branches (i.e. `input-dev`; `permissions`) for development of new features
  + `master` for hosting the production code (**protected branch** now)
+ Breaking: Step 0/1 re-designed (incl. Indicators & Research Steps)
  + imports previous Outcome as Step 0
  + Step 1 pulls from Step 0 based on dropdown (did change yes/no)
  + new YonY submodules & research Substep Components
+ Breaking: Indicators F/P updated to 2020
+ Breaking: variable & property names changed to improve readability
+ [...]

+ new function `isValueInColumn()`
+ `addFileIDtoControl()` now checks for redundant ID in `00-Dashboard`
+ `assignFileOwner(lastname)`; `assignFolderOwner(lastname)` added
+ `assignFileEditors(lastNames)`; `assignFolderEditors(lastNames)`added
  + both sets of methods take string `lastname` or Array string `[lastNames]` and add the email address RDR
  + for the Index, `*Editors()` should take any email

## 2020 February

+ Module: Aggregation - more verbosity; N/P added
+ Module: Data Store extended - imports Scores;
+ overall refactoring & renaming

## 2020 January

+ Module: Spreadsheet Healer extended (dropdowns can be updated / extended); renamed to Health Inspector; reports to `00_Dashboard`
+ Module: Summary Scores extended (column grouping)
+ Scoring Formula overhauled
+ Module: Data Store v2 (long form)

## 2019 December

+ `npm` initialized (`eslint` for Syntax compliance, `clasp`,  `typescript` for Syntax completion)
+ Refactoring: clasp-project isolated and renamed; all code moved to `./src`; json moved to `./json`

## 2019 November 

+ Module: Summary Scores operational
+ New Module: Feedback Request Forms operational
+ New Module: Spreadsheet Healer (Input Sheets: Named Ranges, Labels)
+ New Module: Data Store Layer (beta)

## 2019 October

+ `Pilot` branch
+ refactoring of JSON & DC, SC (naming, cleanup); all modules have an Interface now so that submodules can be called from other Modules (i.e. Scoring submodules can be called from Input Sheet Module)
+ Feature: scoring/feedback can be integrated into DC spreadsheet
+ Module: Aggregation Spreadsheets have standard sheets (scoring logic)
+ Aggregation refactored


## 2019 September

+ added DC Feature: Step-wise grouping of rows
+ added Drive activity parser, that can return atomic activity events (i.e. when a document is changed) for sets of spreadsheets (i.e. time, operation, user). Apart from stats, this could be used for triggering actions.
+ **new Dev Capacity** JS/GAS: added first [`polyfill`](https://developer.mozilla.org/en-US/docs/Glossary/Polyfill): `array.proto.includes`. More to come with further refactoring.
+ implemented layout/formatting options for DC & SC; also: Roboto as default Font Family for better readability; TBD: needs to be refactored into single function
+ introduced coloring based on indicator category
+ **new Feature** `FileID`s of new spreadsheet files are now returned and added to `00_Test` spreadsheet in root of `2019 Back-end testing` folder
+ **new SSOT**: Company JSON is now fetched locally (from a global var in `01_JSON_companies.js`)
+ added Indicator Instructions to `DC`
+ added header formatting patterns; refactoring to `function` TBD
+ split up `data collection` into main and helpers to reduce strain
+ introduced `FeatureRequest` and `bugs` as GitHub Issues (for easier tracking, assigning and elaborating)
+ migrated project to rdresearch
+ implemented `importLocalJSON()` to import JSONs from Google Drive. TODO: needs reliable mechanism for synching JSONs with git [unless we move JSON SSOT to Google Drive which is ok])
+ added output to specific folder functionality

## 2019 August

+ added permission changes module
+ refactored scoring and data collection modules into rough single units of responsibility
+ refactored generic tasks (e.g. importJSONxxx(), nameRange(comp,step,ind)) into helper modules
+ added scoring sheet module
+ enabled NodeJS/clasp + GAS API integration, and full integration with git

## 2019 July

+ initial ground-laying work by tech intern G.W. based on initial requirements from L.G. and prototype architecture from I.S.
+ G.W. created and completed the basic data collection module (99% of the core features), explored the core scoring sheet patterns (i.e. the collection of named ranges for indicator scoring), and created a through documentation of the data collection module

00_mainController.js
// --- // Main Controller // --- //
// --- //  Branch: Development  // --- //

/* global
    centralConfig,
    indicatorsVector,
    filterSingleIndicator,
    companiesVector,
    createSpreadsheetInput,
    createSpreadsheetOutput,
    createFeedbackForms,
    createAggregationOutput,
    createCompanyDataStore,
    processHealthSingleSpreadsheet,
    clearNamedRangesFromCompanySheet,
    openSpreadsheetByID,
    insertSheetIfNotExist,
    addFileIDtoControl,
*/

// global params init (def with initiateGlobalConfig())

/** --- main Prod vs Dev Toggle --- **/
var isProduction = false
/** --- main Prod vs Dev Toggle --- **/

var Config
var IndicatorsObj
var indexPrefix
var filenamePrefix
var filenameSuffix
var rootFolderID
var rootFolderName // "2019 Back-End Dev"
var outputFolderName
var controlSpreadsheetID
var Styles


function initiateGlobalConfig() {
    Config = centralConfig

    IndicatorsObj = indicatorsVector
    // IndicatorsObj = filterSingleIndicator(indicatorsVector, "P11a")

    indexPrefix = Config.indexPrefix
    filenamePrefix = Config.filenamePrefix
    filenameSuffix = isProduction ? Config.filenameSuffixProd : Config.filenameSuffixDev // Dev, "", Debug, QC
    outputFolderName = "2020 Dev Fallback Folder" // Specific folder defined in Main Callers
    rootFolderID = isProduction ? Config.rootFolderIDProd : Config.rootFolderIDDev
    rootFolderName = isProduction ? Config.rootFolderNameProd : Config.rootFolderIDDev
    controlSpreadsheetID = Config.controlSpreadsheetID
    Styles = Config.styles
}

// --- // MAIN CALLERS // --- //

// create Data Collection spreadsheets for all companies

function mainInputSheets() {

    let doClearNamedRanges = false // CAUTION

    initiateGlobalConfig()
    outputFolderName = isProduction ? Config.inputFolderNameProd : Config.inputFolderNameDev
    // filenameSuffix = "" // local override : Dev, "", Debug, QC
    let mainSheetMode = "Input" // for filename
    let useStepsSubset = true // true := use subset
    let useIndicatorSubset = false // true := use subset

    const Companies = companiesVector.companies
        // .slice(0, 0) // on purpose to prevent script from running.
        // .slice(0, 1) //   0 "Alibaba",
        // .slice(1, 2) //   1 "Amazon",
        // .slice(2, 3) //   2 "América Móvil",
        // .slice(3, 4) //   3 "Apple",
        // .slice(4, 5) //   4 "AT&T",
        // .slice(5, 6) //   5 "Axiata",
        // .slice(6, 7) //   6 "Baidu",
        // .slice(7, 8) //   7 "Bharti Airtel",
        // .slice(8, 9) //   8 "Deutsche Telekom",
        // .slice(9, 10) //   9 "Etisalat",
        // .slice(10, 11) //   10 "Facebook",
        // .slice(11, 12) //   11 "Google",
        // .slice(12, 13) //   12 "Kakao",
        // .slice(13, 14) //   13 "Mail.Ru",
        // .slice(14, 15) //   14 "Microsoft",
        // .slice(15, 16) //   15 "MTN",
        // .slice(16, 17) //   16 "Ooredoo",
        // .slice(17, 18) //   17 "Orange",
        // .slice(18, 19) //   18 "Samsung",
        // .slice(19, 20) //   19 "Telefónica",
        // .slice(20, 21) //   20 "Telenor",
        // .slice(21, 22) //   21 "Tencent",
        .slice(22, 23) //   22 "Twitter",
    // .slice(23, 24) //   23 "Verizon Media",
    // .slice(24, 25) //   24 "Vodafone",
    // .slice(25, 26) //   25 "Yandex"

    let fileID

    Companies.forEach(function (Company) {

        fileID = createSpreadsheetInput(useStepsSubset, useIndicatorSubset, Company, filenamePrefix, filenameSuffix, mainSheetMode, doClearNamedRanges)

        addFileIDtoControl(mainSheetMode, Company.label.current, fileID, controlSpreadsheetID)

    })

}

// create Scoring spreadsheets for all companies

function mainScoringSheets() {

    initiateGlobalConfig()
    outputFolderName = "2020 - Dev - Scores"
    let mainSheetMode = "Output"
    let useStepsSubset = false // true := use subset
    let useIndicatorSubset = true // true := use subset

    const Companies = companiesVector.companies
        .slice(1, 9)
    // .slice(0,1) // Amazon
    // .slice(1, 2) // Apple
    // .slice(3,4) //

    let fileID

    Companies.forEach(function (Company) {

        fileID = createSpreadsheetOutput(useStepsSubset, useIndicatorSubset, Company, filenamePrefix, filenameSuffix, mainSheetMode)

        addFileIDtoControl(mainSheetMode, Company.label.current, fileID, controlSpreadsheetID)

    })
}

// create Feedback spreadsheets for all companies

function mainFeedbackSheets() {

    initiateGlobalConfig()
    outputFolderName = "2020 - Dev - Feedback"
    var mainSheetMode = "Feedback"

    var useIndicatorSubset = false // true := use subset

    var Companies = companiesVector.companies
    // .slice(1,2) // Apple

    var fileID

    Companies.forEach(function (Company) {

        fileID = createFeedbackForms(useIndicatorSubset, Company, filenamePrefix, filenameSuffix, mainSheetMode)

        addFileIDtoControl(mainSheetMode, Company.label.current, fileID, controlSpreadsheetID)
    })
}

// create Summary Scores spreadsheets for all companies

function mainAggregationSheets() {

    // filename fragments defined in 
    // Config.summaryParams.spreadsheetName
    initiateGlobalConfig()
    filenameSuffix = "Dev" // DANGER
    outputFolderName = "2020 - Dev - Summary"
    var mainSheetMode = "Summary Scores"

    var useIndicatorSubset = false // true := use subset

    var scoringStepNr = 6

    var Companies = companiesVector.companies
        .slice(1, 9) // no Amazon
    // .slice(1, 3) // for debugging
    // .slice(0,3) // Amazon
    // .slice(1, 2) // Apple
    // .slice(3,4) //

    var fileID = createAggregationOutput(useIndicatorSubset, Companies, filenamePrefix, filenameSuffix, mainSheetMode, scoringStepNr)

    addFileIDtoControl(mainSheetMode, "PROTO", fileID, controlSpreadsheetID)

    Logger.log("added " + mainSheetMode + ";fileID: " + fileID)
}

// create Data Store spreadsheets for all companies

function mainDataStore() {

    var includeWide = false
    initiateGlobalConfig()
    outputFolderName = "2020 - Dev - Data Store"

    filenameSuffix = "Test" // + long or wide is decided in main logic
    // filename fragments defined in 
    // Config.summaryParams.spreadsheetName
    var mainSheetMode = Config.dataStoreParams.fileName

    var useStepsSubset = false // true := use subset
    var useIndySubset = false // true := use subset

    var Companies = companiesVector.companies
        // .slice(1, 9) // exclude Amazon
        // .slice(1, 2) // Apple
        .slice(1, 9)

    var fileID

    Companies.forEach(function (Company) {

        fileID = createCompanyDataStore(useStepsSubset, useIndySubset, Company, filenamePrefix, filenameSuffix, mainSheetMode, includeWide)

        Logger.log("received fileID: " + fileID)
        addFileIDtoControl(mainSheetMode, Company.label.current, fileID, controlSpreadsheetID)

    })
}

// aggregate Input Spreadsheets Health Status (Named Ranges, ..., tbd)

function mainInspectInputSheets() {

    initiateGlobalConfig()
    // IMPORTANT FLAG
    var doRepairs = false // IMPORTANT FLAG

    var mainSheetMode = "Input" // for filename
    filenameSuffix = ""

    var controlSpreadsheet = openSpreadsheetByID(controlSpreadsheetID)
    var ListSheetBroken = insertSheetIfNotExist(controlSpreadsheet, "Input - Broken Refs", true)
    // ListSheetBroken.clear()
    var ListSheetFixed = null

    var Companies = companiesVector.companies
        // .slice(1, 2) // Apple
        .slice(22, 23)

    Companies.forEach(function (Company) {
        processHealthSingleSpreadsheet(ListSheetBroken, ListSheetFixed, Company, filenamePrefix, filenameSuffix, mainSheetMode, doRepairs)
    })

}

// --- // repairing // --- // 
function mainRepairInputSheets() {

    initiateGlobalConfig()
    // IMPORTANT FLAG
    var doRepairs = true // IMPORTANT FLAG

    var mainSheetMode = "Input" // for filename
    filenameSuffix = ""

    var controlSpreadsheet = openSpreadsheetByID(controlSpreadsheetID)
    var ListSheetBroken = insertSheetIfNotExist(controlSpreadsheet, "Input - Broken Refs", true)
    ListSheetBroken.clear()
    var ListSheetFixed

    if (doRepairs) {
        ListSheetFixed = insertSheetIfNotExist(controlSpreadsheet, "Input - Fixed Refs", true)
        // ListSheetFixed.clear()
    } else {
        ListSheetFixed = null
    }

    var Companies = companiesVector.companies
        // .slice(0,3) // Subset #1
        // .slice(3,6) // Subset #2
        // .slice(6,9) // Subset #3
        // .slice(0,1) // Amazon
        // .slice(1, 2) // Apple
        // .slice(2, 3) // Deutsche Telekom
        // .slice(3, 4) // Facebook
        // .slice(4,5) // Google
        // .slice(5,6) // Microsoft
        .slice(6, 7) // Telefonica
    // .slice(7,8) // Twitter
    // .slice(8,9) // Vodafone

    Companies.forEach(function (Company) {
        processHealthSingleSpreadsheet(ListSheetBroken, ListSheetFixed, Company, filenamePrefix, filenameSuffix, mainSheetMode, doRepairs)
    })

}

// --- // USE WISELY // --- // 
/* function mainClearAllNamedRanges() {

    initiateGlobalConfig()
    var mainSheetMode = "Input" // for filename

    var Companies = companiesVector.companies
        .slice(0, 0)
    // .slice(0,1) // Amazon
    // .slice(1, 2) // Apple
    // .slice(2,3) // Deutsche Telekom
    // .slice(3,4) // Facebook
    // .slice(4,5) // Google
    // .slice(5,6) // Microsoft
    // .slice(6,7) // Telefonica
    // .slice(7,8) // Twitter
    // .slice(8,9) // Vodafone

    Companies.forEach(function (Company) {
        clearNamedRangesFromCompanySheet(Company, filenamePrefix, filenameSuffix, mainSheetMode)
    })

} */
